pkg/controllers/management/rkeworkerupgrader/node.go

```diff
@@ -187,3 +188,47 @@ func workerOnly(roles []string) bool {
 func ignoreNode(node *v3.Node) bool {
 	return node.Status.NodeLabels != nil && node.Status.NodeLabels[ignoreKey] == ignoreValue
 }
+
+func CanUpgrade(nodes []*v3.Node, upgradeStrategy *v3.NodeUpgradeStrategy) error {
+	var (
+		controlReady, controlNotReady, workerReady, workerNotReady int
+	)
+	for _, node := range nodes {
+		if node.Status.NodeConfig == nil {
+			continue
+		}
+		if slice.ContainsString(node.Status.NodeConfig.Role, rkeservices.ControlRole) {
+			if nodehelper.IsMachineReady(node) {
+				controlReady++
+			} else {
+				controlNotReady++
+			}
+			continue
+		}
+		if slice.ContainsString(node.Status.NodeConfig.Role, rkeservices.ETCDRole) {
+			continue
```

Comment:
If a box can be all 3 roles, wouldn't this logic skip some ? IE: box-a roles are `[ctrl, etcd, worker]` and box-b roles are `[worker]`, this would only check worker status on box-b.

Reply:
This would count box-a as a `ctrl` node and box-b as `worker`. To clarify, if a node has all 3 roles, it's only considered as control node. I will rename the worker to `workerOnly`, does that answer your example?

Reply:
yes, as long as we are ok with that scenario, and scenario where box is `[etcd, worker]`